### ISSUE TITLE
Implement session timeout on brokered Solr instances

### ISSUE THREAD

**Issue by FuhuXia** [GSA/data.gov#3273]
_Please keep any sensitive details in [Google Drive](https://docs.google.com/document/d/1rW0VOzfCrjXOI0O1gTAOm225_fTbu5Cjrg7yJTYYBBg/edit#heading=h.ykwjreds5oap)._

**Date of report:** 2021-05-26
**Severity:** Moderate
**Due date:** 2021-08-26

_Due date is based on severity and described in RA-5. 15-days for Critical, 30-days for High, and 90-days for Moderate and lower._

- [ ] Analysis has been performed and an issue has been linked to address other occurrences for this class of vulnerability\* (_link_)

_\* When a finding is identified, we create two issues. One to address the specific instance identified in the report. The other is to identify and address all other occurrences of this vulnerability within the application._

**Comment by mogul** [comment]
[Example of an Angular timeout implementation](https://github.com/jackjamieson/ng-timeout); we could contribute a PR that adds this upstream.

**Comment by mogul** [comment]
A potential avenue for handling this at the proxy layer [suggested in Slack](https://gsa-tts.slack.com/archives/CG8SNAEQM/p1630356494005900?thread_ts=1630353194.003800&cid=CG8SNAEQM):
> How about a scheme like this, which is heavily dependent on app behavior/implementation:
> - all traffic flows through reverse proxy
> - leave the app authn/logout URL unrestricted
> - look for the app-level Set-Cookie to set your own Module cookie (perhaps encrypted with session expiration, or just pointing to a persistent session)
> - when your cookie is present but is expired, redirect to the app logout URL (rewrite the HTTP response headers)
> - if your cookie is not present, redirect to the app authn URL (rewrite the HTTP response headers)
> - if both cookies are present, and the module cookie is not expired, allow the request through

**Comment by mogul** [comment]
[Another useful find](https://github.com/bungle/lua-resty-session), note in particular `session.cookie.idletime`. Possibly a reason to move Solr auth back to nginx-ingress annotations, though I have no idea how we'd handle logout in that case. 🤔

**Comment by mogul** [comment]
Actually I think this just makes it straightforward to implement the idea above.

Note that nginx-ingresses can be [annotated with nginx.conf snippets](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#configuration-snippet) or [customized with lua plugins](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#plugins).

**Comment by mogul** [comment]
[external_auth](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#external-authentication) might be an option too.

**Comment by nickumia-reisys** [comment]
After all of the spin-up work, I can understand your comments better 😄.  Supporting login/logout and session timeouts were created a two separate issues.  In more advanced authentication systems, we would be able to group these two things together.  For simplicity, I propose keeping them separate.  What if we had an unauthenticated session cookie tracked by NGINX that enabled support of timeouts.  So, for a complete valid session, the NGINX cookie has to not be expired and the user needs to be logged into Solr through it's native authentication support.  

After typing the above, I realized that the user would still be logged into Solr even if the NGINX cookie expires.  (And this is where the idea breaks down and may not work..)

If the session cookie expires, we can maybe add a script/option/something that will run through the logout on Solr Admin UI.

**Comment by mogul** [comment]
Yup! I think you're coming around to the implementation suggested above: https://github.com/GSA/datagov-deploy/issues/3273#issuecomment-908713765

**Comment by mogul** [comment]
I have a suspicion that we can use `external_auth` to handle both auth and idle timeout for the Solr instance... where the Solr instance itself is the external auth provider, based on the work we just did! We'll have to experiment. Happily as it happens that method may also be the easiest to try out.

**Comment by nickumia-reisys** [comment]
The logout could be changing the password to force logout and then changing it back to allow the same password to be used again.

**Comment by mogul** [comment]
Interesting idea but the admin password necessary to do that wouldn't be available to nginx in that context.

**Comment by jbrown-xentity** [comment]
Should bring up at CKAN dev meetings, ask if later version of SOLR has been tested (currently 6.6.6 is suggested).
